### PR TITLE
Fix error: conflicting types for built-in function 'execl'; expected 'int(const char *, const char *, ...)

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
prototype here: https://pubs.opengroup.org/onlinepubs/9699919799/functions/exec.html

## Testing
Pass CI

### CODE REVIEW

File: libs/libc/unistd/lib_execl.c [hunk 1]
```diff
@@ -113,67 +113,61 @@
  *
  ****************************************************************************/
 
-int execl(FAR const char *path, ...)
+int execl(FAR const char *path, FAR const char *arg0, ...)
 {
   FAR char **argv = (FAR char **)NULL;
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:118]:
```suggestion
  FAR char * const *argv = NULL;
```

File: libs/libc/unistd/lib_execl.c [hunk 2]
```diff
@@ -113,67 +113,61 @@
...
+      arg = va_arg(ap, FAR char *);
     }
-  while (arg);
 
   va_end(ap);
 
   /* Allocate a temporary argv[] array */
 
-  if (nargs > 0)
+  argv = (FAR char **)lib_malloc((nargs + 1) * sizeof(FAR char *));
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:152]:
```suggestion
  argv = (FAR char * const *)lib_malloc((nargs + 1) * sizeof(FAR char *));
```

File: libs/libc/unistd/lib_execl.c [hunk 3]
```diff
@@ -113,67 +113,61 @@
...
+  if (argv == (FAR char **)NULL)
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:153]:
```suggestion
  if (argv == NULL)
```

File: libs/libc/unistd/lib_execl.c [hunk 4]
```diff
@@ -113,67 +113,61 @@
...
-        {
-          set_errno(ENOMEM);
-          return ERROR;
-        }
+      set_errno(ENOMEM);
+      return ERROR;
+    }
 
-      /* Collect the arguments into the argv[] array */
+  argv[0] = (FAR char *)arg0;
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:159]:
```suggestion
  argv[0] = arg0;
```

File: libs/libc/unistd/lib_execl.c [hunk 5]
```diff
@@ -113,67 +113,61 @@
...
-          argv[argc] = va_arg(ap, FAR char *);
-        }
+  /* Collect the arguments into the argv[] array */
 
-      argv[nargs] = NULL;
-      va_end(ap);
+  va_start(ap, arg0);
+  for (argc = 1; argc <= nargs; argc++)
+    {
+      argv[argc] = va_arg(ap, FAR char *);
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:166]:
```suggestion
      argv[argc] = va_arg(ap, FAR const char *);
```

File: libs/libc/unistd/lib_execl.c [hunk 6]
```diff
@@ -113,67 +113,61 @@
  *
  ****************************************************************************/
 
-int execl(FAR const char *path, ...)
+int execl(FAR const char *path, FAR const char *arg0, ...)
 {
   FAR char **argv = (FAR char **)NULL;
-  FAR char *arg;
+  FAR char *arg = (FAR char *)arg0;
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:119]:
```suggestion
  FAR const char *arg = arg0;
```

File: libs/libc/unistd/lib_execl.c [hunk 7]
```diff
@@ -113,67 +113,61 @@
...
-              set_errno(E2BIG);
-              va_end(ap);
-              return ERROR;
-            }
+          set_errno(E2BIG);
+          va_end(ap);
+          return ERROR;
         }
+
+      arg = va_arg(ap, FAR char *);
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:145]:
```suggestion
      arg = va_arg(ap, FAR const char *);
```

File: libs/libc/unistd/lib_execl.c [hunk 8]
```diff
@@ -113,67 +113,61 @@
...
+  for (argc = 1; argc <= nargs; argc++)
+    {
+      argv[argc] = va_arg(ap, FAR char *);
     }
 
+  va_end(ap);
+
   /* Then let execv() do the real work */
 
   ret = execv(path, (FAR char * const *)argv);
```
Comment (pkarashchenko) [inline libs/libc/unistd/lib_execl.c:173]:
```suggestion
  ret = execv(path, argv);
```

**Review by pkarashchenko — COMMENTED** [review]
I'm not sure if everything that I commented will work, but I hope it should.

File: libs/libc/unistd/lib_execl.c (hunk 1 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done.

File: libs/libc/unistd/lib_execl.c (hunk 2 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done.

File: libs/libc/unistd/lib_execl.c (hunk 3 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done.

File: libs/libc/unistd/lib_execl.c (hunk 4 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done.

File: libs/libc/unistd/lib_execl.c (hunk 6 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done.

File: libs/libc/unistd/lib_execl.c (hunk 8 above)
Comment (xiaoxiang781216) [reply libs/libc/unistd/lib_execl.c]: Done,

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Comment by pkarashchenko** [comment]
@xiaoxiang781216 so the changes with `const` didn't work?

**Comment by pkarashchenko** [comment]
> > @xiaoxiang781216 so the changes with `const` didn't work?
> 
> Yes, the compiler report: can't modify const argv.

Maybe I put `const` in a wrong place. I think it should be `FAR const char **argv = NULL;` and not `FAR char * const *argv = NULL;` as I initially suggested. We need a pointer to a pointer that points a constant data and `FAR char * const *argv = NULL;` means that `argv[0] =` is illegal. My mistake

**Comment by xiaoxiang781216** [comment]
> > > @xiaoxiang781216 so the changes with `const` didn't work?
> > 
> > 
> > Yes, the compiler report: can't modify const argv.
> 
> Maybe I put `const` in a wrong place. I think it should be `FAR const char **argv = NULL;` and not `FAR char * const *argv = NULL;`

If so, we require cast in two place: execv and lib_free.

> as I initially suggested. We need a pointer to a pointer that points a constant data and `FAR char * const *argv = NULL;` means that `argv[0] =` is illegal. My mistake

I has tried all possible combination(const char **, char * const *, const char * const), char ** is the best option since it require the less cast.

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**